src/main/scala/Core/Top.scala

```diff
@@ -35,23 +37,25 @@ class VerilatorTop(cfg: FlexpretConfiguration) extends AbstractTop(cfg) {
     core.io.int_exts.foreach(_ := false.B)
 
     // Catch termination from core
-    when(core.io.host.to_host === "hdeaddead".U) {
-        printf(cf"SUCCESS: FlexPRET terminated execution\n")
-        io.stop := true.B
-    }
-
-    // Catch abort from core
-    when(core.io.host.to_host === "hdeadbeef".U) {
-        printf(cf"ERROR: FlexPRET aborted simulation\n")
-        assert(false.B, "Emulation stopped")
-    }
-
-    // Handle printfss
-    when(core.io.host.to_host === "hbaaabaaa".U) {
-      regPrintNext := true.B
-    }.elsewhen(regPrintNext) {
-      printf(cf"${core.io.host.to_host}\n")
-      regPrintNext := false.B
+    for (tid <- 0 until cfg.threads) {
+        when(core.io.host.to_host(tid) === "hdeaddead".U) {
+            io.stop := true.B
+        }
+
+        // Catch abort from core
+        when(core.io.host.to_host(tid) === "hdeadbeef".U) {
+            io.stop := true.B
+        }
```

Comment:
Wasnt the `_fp_abort` function removed?

Reply:
The _fp_abort function still exists, but is handled by verilator. But I guess the CPU still needs to set the io.stop signal when it gets it? Hmm, or maybe not actually.

Reply:
I guess you are right, so forget my comment here.  I dont think there is another way of terminating the verilator emulation besides driving the io.stop signal to high?

Reply:
Hmm yes I think it is possible by just comparing the to_host signals to the 0xdeaddead and 0xdeadbeef values directly, without the first if-statement here. It might impact performance of the emulator, however, having to check every single thread for those two stop values each cycle, as opposed to just checking the io_stop signal each cycle. But I have no idea by how much it will impact the performance of the emulator.

```
    if (top->io_stop == 1) {
      // Check for abort signals from FlexPRET and propagate the exit code
      // by returning it from the emulator
      bool unknown_reason = true;
      for (int i = 0; i < NUM_THREADS; i++) {
        const uint32_t to_host = get_to_host(i, top);
        if (to_host == 0xdeaddead) {
          exitcode = EXIT_SUCCESS;
          unknown_reason = false;
        } else if (to_host == 0xdeadbeef) {
          exitcode = EXIT_FAILURE;
          unknown_reason = false;
        }
      }

      if (unknown_reason) {
        printf("%s: Exit due to unknown reason\n", argv[0]);
        exitcode = EXIT_FAILURE;
      }

      break;
    }
```